src/main/kotlin/ecommerce/interceptor/AuthInterceptor.kt

```diff
@@ -0,0 +1,106 @@
+        const val USER_EMAIL_ATTRIBUTE = "userEmail"
+        const val USER_NAME_ATTRIBUTE = "userName"
+        private const val BEARER_PREFIX = "Bearer "
+        private const val ADMIN_PATH_PREFIX = "/admin/"
+        private const val ADMIN_ROLE = "ADMIN"
+    }
+
+    override fun preHandle(
+        request: HttpServletRequest,
+        response: HttpServletResponse,
+        handler: Any,
+    ): Boolean {
+        return try {
+            val token =
+                extractToken(request) ?: return sendUnauthorizedResponse(
+                    response, "Missing or invalid Authorization header",
+                )
+
+            val claims =
+                tokenService.validateToken(token) ?: return sendUnauthorizedResponse(
+                    response, "Invalid or expired token",
+                )
+
+            val userId =
+                extractUserId(claims) ?: return sendUnauthorizedResponse(
+                    response, "Invalid token payload",
+                )
+
+            storeUserAttributes(request, claims, userId)
+
+            if (requiresAdminAccess(request) && !hasAdminRole(claims)) {
+                return sendForbiddenResponse(response, "Admin access required")
+            }
+
+            true
+        } catch (e: Exception) {
+            sendUnauthorizedResponse(response, "Authentication error: ${e.message}")
+        }
+    }
+
+    private fun extractToken(request: HttpServletRequest): String? {
+        val authHeader = request.getHeader("Authorization")
+        return if (authHeader?.startsWith(BEARER_PREFIX) == true) {
+            authHeader.substring(BEARER_PREFIX.length)
+        } else {
+            null
+        }
+    }
+
+    private fun extractUserId(claims: io.jsonwebtoken.Claims): Long? {
+        return claims.subject?.toLongOrNull()
+    }
+
+    private fun storeUserAttributes(
+        request: HttpServletRequest,
+        claims: io.jsonwebtoken.Claims,
+        userId: Long,
+    ) {
+        request.setAttribute(USER_ID_ATTRIBUTE, userId)
+        request.setAttribute(USER_ROLE_ATTRIBUTE, claims["role"] as? String)
+        request.setAttribute(USER_EMAIL_ATTRIBUTE, claims["email"] as? String)
+        request.setAttribute(USER_NAME_ATTRIBUTE, claims["name"] as? String)
+    }
+
+    private fun requiresAdminAccess(request: HttpServletRequest): Boolean {
+        val uri = request.requestURI
+        return uri == "/admin" || uri.startsWith(ADMIN_PATH_PREFIX)
+    }
+
+    private fun hasAdminRole(claims: io.jsonwebtoken.Claims): Boolean {
+        return claims["role"] as? String == ADMIN_ROLE
+    }
+
+    private fun sendUnauthorizedResponse(
+        response: HttpServletResponse,
+        message: String,
+    ): Boolean {
+        response.status = HttpServletResponse.SC_UNAUTHORIZED
+        response.contentType = "application/json"
+        response.writer.write("""{"error": "Unauthorized", "message": "$message"}""")
+        return false
+    }
```

Comment:
Instead of writing directly, try throwing an exception so that `ControllerAdvice` can handle it. This is possible in an Interceptor!